PR 4 of n - Add pathway analysis GSVA notebook (#281)
AlexsLemonade/training-modules: pathway-analysis/03-gene_set_variation_analysis.Rmd

We have VST transformed RNA-seq data, annotated with gene symbols, that has been collapsed such that there are no duplicated gene identifiers (see `setup`).

```{r input_file}
rnaseq_file <- file.path(input_dir, "medulloblastoma_vst_collapsed.tsv")
```

#### Output

```{r output_file, live = TRUE}
gsva_results_file <- file.path(output_dir, "medulloblastoma_gsva_results.tsv")
```

## Gene sets

The function that we will use to run GSVA wants the gene sets to be in a list, rather than a tidy data frame that we used with `clusterProfiler` (although it does accept other formats).

We're going to take this opportunity to introduce a different format that gene sets are often distributed in called [GMT (Gene Matrix Transposed)](https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GMT:_Gene_Matrix_Transposed_file_format_.28.2A.gmt.29).

We're going to read in the Hallmark collection file from [MSigDB](https://www.gsea-msigdb.org/gsea/msigdb/index.jsp), rather than using `msigdbr` like we did in earlier notebooks.

The RNA-seq data uses gene symbols, so we need gene sets that use gene symbols, too.

```{r gmt}
# R can often read in data from a URL

> hallmarks_url <- "https://data.broadinstitute.org/gsea-msigdb/msigdb/release/7.1/h.all.v7.1.symbols.gmt"

Author: Honestly I had to look at refine.bio examples to find it 😂

                     # Compute Gaussian-distributed scores
                     mx.diff = TRUE,
                     # Use 4 cores
                     parallel.sz = 4)
```

**Note: the `gsva()` documentation says we can use `kcdf = "Gaussian"` if we had RNA-seq log-CPMs, log-RPKMs or log-TPMs, but we would use `kcdf = "Poisson"` on integer counts.**

```{r gsva_peek}
# Let's explore what the output of gsva() looks like
gsva_results[1:5, 1:5]
```

### A note on gene set size
 
Often the scores of gene set enrichment methods are not comparable between gene sets of different sizes.
Let's do an experiment using randomly generated gene sets to explore this idea a bit more.

```{r generate_random_genesets}
# Use all the gene symbols in the dataset as the pool of possible genes
all_genes <- rownames(rnaseq_mat)

# Set a seed for reproducibility
set.seed(2020)

# Make a list of integers that indicate the random gene set sizes
gene_set_size <- list(15, 25, 50, 100, 250, 500)

# Generate 100 random gene sets of each size
random_gene_sets <- 
  purrr::map(
    1:100,  # Repeat the function 100 times
    ~ purrr::map(
        # For the list of gene set sizes

>         gene_set_size, 

Author: This makes sense!

                      values_to = "gsva_score") %>%
  # Remove the .version added by make.names()
  dplyr::mutate(gene_set = stringr::str_remove(gene_set, "\\..*")) %>%
  # Add a column that keeps track of the gene set size
  dplyr::mutate(gene_set_size = stringr::word(gene_set, 2, sep = "_")) %>%
  # We want to plot smallest no. genes -> largest no. genes
  dplyr::mutate(gene_set_size = factor(gene_set_size, 
                                       levels = c(15, 25, 50, 100, 250, 500)))

```

Let's make a violin plot so we can look at the distribution of scores by gene set size.

```{r random_violin}
# Violin plot comparing GSVA scores of different random gene set sizes
random_long_df %>%
  ggplot2::ggplot(ggplot2::aes(x = gene_set_size, 
                               y = gsva_score)) +
  # Make a violin plot
  ggplot2::geom_violin() +
  # Add a point with the mean value
  ggplot2::stat_summary(
    geom = "point",
    fun.y = "mean"
  ) +
  # Flip the axes
  ggplot2::coord_flip() +
  ggplot2::labs(title = "Random Gene Set GSVA Scores",
                x = "gene set size",
                y = "GSVA score") +
  ggplot2::theme_bw()
```

> What do you notice about these distributions?

Reviewer: I want to suggest something about getting to significance through randomization here. Is that beyond the scope?
Does GSVA implement something like this?

```suggestion
What do you notice about these distributions?
How might you use this information to inform your interpretation of GSVA scores?
```

Author: I don't think GSVA implements anything like this. The way I think about "significance" of pathway scores in this vein is testing for differential expression if you happen to have group information, rather than an individual sample's score being different from a null distribution.  Quoting from the GSVA paper:

> Although the GSVA algorithm itself does not evaluate statistical significance for the enrichment of gene sets, significance with respect to a phenotype can be easily evaluated using conventional statistical models. Likewise, false discovery rates can be estimated by permuting the sample labels (Methods). We make no general prescription for thresholds of significance or false discovery, as these choices are highly context dependent and may vary according to each experiment.

I guess my point was more that you may want to make your scores more comparable (maybe scaling would be enough, without generating random gene sets...) if you're going to try to compare between pathways. Was that clear? It is after 5pm on Friday 😅

Author: I can add your suggested sentence either way!

### How can you use these scores?

If you did have groups information for your samples, you could test for pathway scores differences between groups.

> Here's [an example](https://htmlpreview.github.io/?https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/9b44bf1c186b3126b16dbe5b87756b3eae3feec2/analyses/gene-set-enrichment-analysis/02-model-gsea.nb.html) of that from the OpenPBTA project itself! 

Reviewer: (don't look at the file name)

Author: 🙈

---
title: "Pathway analysis: Gene Set Variation Analysis (GSVA)"
output: 
  html_notebook:
    toc: true
    toc_float: true
author: CCDL for ALSF
date: 2020
---

> So far every pathway analysis method we've covered relies on some information about groups in our data.

Reviewer: ```suggestion
So far every pathway analysis method we've covered relies on some information about groups of samples in our data.
```

**Figure 1 from [Hänzelmann, Castelo, and Guinney. (2013)](https://doi.org/10.1186/1471-2105-14-7).**

You may notice that GSVA has some commonalities with GSEA.
Rather than ranking genes based on some statistic _we_ selected ahead of time, GSVA fits a model and ranks genes based on their expression level relative to the sample distribution.
This is a way of asking if a gene _i_ is highly or lowly expressed in a sample _j_ in the context of this experiment and ranking accordingly ([Hänzelmann, Castelo, and Guinney. 2013](https://doi.org/10.1186/1471-2105-14-7)).
The pathway-level score calculated is a way of asking how genes _within_ a gene set vary as compared to genes that are _outside_ of that gene set ([Malhotra. 2018](https://towardsdatascience.com/decoding-gene-set-variation-analysis-8193a0cfda3)). 
(This is sometimes called a competitive test in gene set enrichment literature.)

> The inituition here is that we will get a pathway-level score that indicates if genes in a pathway that act concordantly in one direction (overexpressed or underexpressed relative to the overall population) ([Hänzelmann, Castelo, and Guinney. 2013](https://doi.org/10.1186/1471-2105-14-7)).

Reviewer: Spelling fix and clarification (to my understanding, but correct me if I am wrong.)
```suggestion
The intuition here is that we will get pathway-level scores for each sample that indicate if genes in a pathway vary concordantly in one direction (overexpressed or underexpressed relative to the overall population) ([Hänzelmann, Castelo, and Guinney. 2013](https://doi.org/10.1186/1471-2105-14-7)).
```
